gan/gan_deepspeed_train.py

```diff
@@ -0,0 +1,177 @@
+    create_folder(args.outf)
+    set_seed(args.manualSeed)
+    cudnn.benchmark = True
+    dataset, nc = get_dataset(args)
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=args.batchSize, shuffle=True, num_workers=int(args.workers))
+    torch.cuda.set_device(args.local_rank)
+    device = torch.device("cuda", args.local_rank) #torch.device("cuda:0" if args.cuda else "cpu")
+    ngpu = 0
+    nz = int(args.nz)
+    ngf = int(args.ngf)
+    ndf = int(args.ndf)
+
+    netG = Generator(ngpu, ngf, nc, nz).to(device)
+    netG.apply(weights_init)
+    if args.netG != '':
+        netG.load_state_dict(torch.load(args.netG))
+
+    netD = Discriminator(ngpu, ndf, nc).to(device)
+    netD.apply(weights_init)
+    if args.netD != '':
+        netD.load_state_dict(torch.load(args.netD))
+
+    criterion = nn.BCELoss()
+
+    fixed_noise = torch.randn(args.batchSize, nz, 1, 1, device=device)
+    real_label = 1
+    fake_label = 0
+
+    # setup optimizer
+    optimizerD = torch.optim.Adam(netD.parameters(), lr=args.lr, betas=(args.beta1, 0.999))
+    optimizerG = torch.optim.Adam(netG.parameters(), lr=args.lr, betas=(args.beta1, 0.999))
+
+    model_engineD, optimizerD, _, _ = deepspeed.initialize(args=args, model=netD, model_parameters=netD.parameters(), optimizer=optimizerD)
+    model_engineG, optimizerG, _, _ = deepspeed.initialize(args=args, model=netG, model_parameters=netG.parameters(), optimizer=optimizerG)
+
+    for epoch in range(args.epochs):
+        for i, data in enumerate(dataloader, 0):
+            ############################
+            # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
+            ###########################
+            # train with real
+            netD.zero_grad()
+            real_cpu = data[0].to(device)
+            batch_size = real_cpu.size(0)
+            label = torch.full((batch_size,), real_label, dtype=real_cpu.dtype, device=device)
+            output = netD(real_cpu)
+            errD_real = criterion(output, label)
+            model_engineD.backward(errD_real)
+            D_x = output.mean().item()
+
+            # train with fake
+            noise = torch.randn(batch_size, nz, 1, 1, device=device)
+            fake = netG(noise)
+            label.fill_(fake_label)
+            output = netD(fake.detach())
+            errD_fake = criterion(output, label)
+            model_engineD.backward(errD_fake)
+            D_G_z1 = output.mean().item()
+            errD = errD_real + errD_fake
+            #optimizerD.step() # alternative (equivalent) step
+            model_engineD.step()
+
+            ############################
+            # (2) Update G network: maximize log(D(G(z)))
+            ###########################
+            netG.zero_grad()
+            label.fill_(real_label)  # fake labels are real for generator cost
+            output = netD(fake)
+            errG = criterion(output, label)
+            model_engineG.backward(errG)

```

Comment:
I have one question here, what will happen in this backward()? will both G and D generate their gradients? since D already finish its step(), will its parameters be impacted by this?

Reply:
Yes, both G and D will generate gradients here. This will affect D only if there is gradient accumulation across iterations (in which case requires_grad of D's parameters can be set to false on the client side before model_engineG.backward(errG)). However, in this example, netD.zero_grad() clears the gradients in D (due to the backward on G) before updating D.